client/consensus/babe/src/authorship.rs

```diff
@@ -241,16 +240,26 @@ fn claim_primary_slot(
 		// be empty.  Therefore, this division in `calculate_threshold` is safe.
 		let threshold = super::authorship::calculate_primary_threshold(c, authorities, *authority_index);
 
-		let pre_digest = get_keypair(pair)
-			.vrf_sign_after_check(transcript, |inout| super::authorship::check_primary_threshold(inout, threshold))
-			.map(|s| {
-				PreDigest::Primary(PrimaryPreDigest {
-					slot_number,
-					vrf_output: VRFOutput(s.0.to_output()),
-					vrf_proof: VRFProof(s.1),
-					authority_index: *authority_index as u32,
-				})
-			});
+		let (output, proof) = keystore.read().vrf_sign(
+			pair.as_ref(), &BABE_ENGINE_ID, randomness, slot_number, *epoch_index,
+		);
+		let proof = schnorrkel::vrf::VRFProof::from_bytes(&proof).ok()?;
```

Comment:
This could return None from within the loop which removes the possibility of iterating over all "pairs". I think the better alternative is that if this is None, we just ignore and move to the next... opinions?

Reply:
You want the remote signer to take the `randomness` and `epoch_index`, and a range for 					`slot_number`, repeatedly call `make_transcript` and `vrf_sign_after_check`, and  return roughly `Vec<(slot_number,VRFOutput,VRFProof)>` whenever the threshold check passes?  

Yes, I'd think doing that fits the threat model well.

Reply:
You lost me at the question mark :grin:. Are you suggesting what i should do? or are you pointing out what i am doing is wrong?

Reply:
Yes it's a suggestion.  I've edited the comment.  :)

Reply:
Understood... your meaning is to batch the vrf signing call to replace the loop here with a single call:
https://github.com/paritytech/substrate/blob/6ec597649b3112e4e809afd1b36888b2872275b9/client/consensus/babe/rpc/src/lib.rs#L142-L158

correct?

Reply:
Another question: 

> and then repeatedly calls make_transcript and vrf_sign_after_check
So you think that `calculate_primary_threshold` and `check_primary_threshold` can be done on the server side instead of the client side (which my code is doing atm)?

Reply:
Yes, you query for either the next or for a batch of VRF outputs.  I'd call `calculate_primary_threshold` in the node and just pass the threshold to the remote signer.  

We address compromised node setting better by doing transcript construction inside the signer, which restricts the signable messages, but the same cannot be said for `calculate_primary_threshold`.